### ISSUE TITLE
[AICA5] ernie_doc 如何使用长文本微调，有没有具体的示例

### ISSUE THREAD

**Issue by mymoongit** [PaddlePaddle/PaddleNLP#1217]
- 版本、环境信息
1）PaddleNLP 2.1.0，PaddlePaddle2.1.3
2）系统环境：Windows10，python 3.7
- inputs = tokenizer("欢迎使用百度飞桨！")
ernie_doc 怎样处理长文本输入，如输入长度3000+的文本，

**Comment by wawltor** [comment]
您好，https://github.com/PaddlePaddle/PaddleNLP/tree/develop/examples/language_model/ernie-doc 这里目录下有一些关于文本分类，信息抽取，文本匹配相关的长文本finetune的示例，可以了解一下

**Comment by joey12300** [comment]
@mymoongit ERNIE-Doc在不同任务下对原始数据的预处理也不同，详细可参考该数据处理脚本https://github.com/PaddlePaddle/PaddleNLP/blob/develop/examples/language_model/ernie-doc/data.py

**Comment by mymoongit** [comment]
@wawltor ernie-doc 分类中，
logits, memories = model(input_ids, memories, token_type_ids, position_ids, attn_mask)

memories 放在不同的batch 循环，具体作用是什么；上一批样本的memories 跟当前的样本有什么关系呢

**Comment by joey12300** [comment]
@mymoongit ernie-doc是一种基于循环机制的transformer，会将一个长文本分割成若干个等长、有序的短文本，然后依次读入这些短文本。为了关联起不同批次的短文本，当前批次的训练除了读入当前批次的输入（input_ids）以外，还要融入上一批次的输出信息（memories）。

更多细节可参考文章：ERNIE-Doc: A Retrospective Long-Document Modeling Transformer

**Comment by mymoongit** [comment]
老师您好，感谢回复！

ernie-doc 数据预处理，过于复杂；我想用它处理长文本分类；

data.py 跟踪每个批次的qid
for batch_records in self._prepare_batch_data(sample_batch):
    print(np.array(batch_records[5]).flatten())
    print()
    yield batch_records

显示每批次的qid如下：
[0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19]

[20 21  2 22 23 24  6 25 26  9 27 11 12 13 28 29 30 31 18 19]

[32 33  2 22 34 24  6 35 36  9 27 11 37 13 28 38 30 31 18 19]

[32 39  2 40 34 24  6 35 36  9 27 41 42 13 28 43 30 31 18 19]

[32 44  2 40 34 24  6 35 36  9 27 45 46 13 28 47 30 31 18 19]

[32 44  2 48 34 24  6 35 36  9 27 45 46 13 28 49 30 50 18 19]

run_classifier.py 分类文件中，memories 一直在循环中，也就是上一批次的输出会作为下一批次的输入，
但不同批次间对应的qid可能是不一样的，当前文档可加入其他文档的memories么？
# Copy the memory
memories = create_memory()

tic_train = time.time()
for epoch in range(args.epochs):
# train_ds_iter.shuffle_sample()
train_dataloader.set_batch_generator(train_ds_iter, paddle.get_device())

for step, batch in enumerate(train_dataloader, start=1):

        global_steps += 1
        input_ids, position_ids, token_type_ids, attn_mask, labels, qids, gather_idx, need_cal_loss = batch

        logits, memories = model(input_ids, memories, token_type_ids, position_ids, attn_mask)

**Comment by joey12300** [comment]
@mymoongit 感谢你的提问！ ernie_doc的循环Transformer机制最早是由TransformerXL提出的。TransformerXL开源的代码存在你所说的跨qid的memories传递，即`memories 一直在循环中，也就是上一批次的输出会作为下一批次的输入，
但不同批次间对应的qid可能是不一样的`。作者为了方便对比ernie_doc和TransformerXL之间的效果，也直接复用了TransformerXL的跨qid的memories传递方法。

**Comment by mymoongit** [comment]
@joey12300 感谢回答，了解了，批处理中大多qid是一致的，少部分不同，作为引入的噪声或是其他考虑；
但，qid 一致的话也是可行的，是吧；如遇到qid与当前不一致的地方，对应memory 可以阻断或替换为初始memory,
